### PR TITLE
[Merged by Bors] - feat(Data/Nat/Digits): sub_one_mul_sum_div_pow_eq_sub_sum_digits

### PR DESCRIPTION
**Pull request by mo271**
- [x] depends on: #6448
- [x] depends on: #6493

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
To be used in #5803

File: Mathlib/Data/Nat/Digits.lean [hunk 1]
```diff
@@ -516,6 +517,59 @@ lemma self_div_pow_eq_ofDigits_drop (i n : ℕ) (h : 2 ≤ p):
     (fun l hl ↦ digits_lt_base h hl)
   exact (ofDigits_digits p n).symm
 
+open BigOperators
```
Comment (ocfnash) [inline Mathlib/Data/Nat/Digits.lean:520]: If you also open `Finset` then you'll be able to drop a bunch of `Finset.` prefixes below.
```suggestion
open BigOperators Finset
```

File: Mathlib/Data/Nat/Digits.lean [hunk 2]
```diff
@@ -516,6 +517,59 @@ lemma self_div_pow_eq_ofDigits_drop (i n : ℕ) (h : 2 ≤ p):
...
+
+theorem sub_one_mul_sum_div_pow_eq_sub_sum_digits' (L : List ℕ) (w₁ : ∀ (l : ℕ), l ∈ L → l < p)
+    (w₂ : ∀ (h : L ≠ []), L.getLast h ≠ 0) : (p - 1) * ∑ i in Finset.range L.length,
+    (ofDigits p L) / p ^ i.succ = (ofDigits p L) - L.sum := by
```
Comment (ocfnash) [inline Mathlib/Data/Nat/Digits.lean:524]: I think a slightly better statement (which depends on the `open Finset` above) is:
```suggestion
theorem sub_one_mul_sum_div_pow_eq_sub_sum_digits
    (L : List ℕ) {h} (h_ne_zero : L.getLast h ≠ 0) (h_lt : ∀ l ∈ L, l < p) :
    (p - 1) * ∑ i in range L.length, (ofDigits p L) / p ^ i.succ = (ofDigits p L) - L.sum := by
```

File: Mathlib/Data/Nat/Digits.lean [hunk 3]
```diff
@@ -516,6 +517,59 @@ lemma self_div_pow_eq_ofDigits_drop (i n : ℕ) (h : 2 ≤ p):
...
+  obtain h | h | h : 1 < p ∨ 1 = p ∨ p < 1 := trichotomous 1 p
```
Comment (ocfnash) [inline Mathlib/Data/Nat/Digits.lean:525]: If you use this trick then you won't have to rewrite by `← h` in the middle case below.
```suggestion
  obtain h | rfl | h : 1 < p ∨ 1 = p ∨ p < 1 := trichotomous 1 p
```

Comment (ocfnash) [reply Mathlib/Data/Nat/Digits.lean:525]: (same comment applies in the proof of the corollary below)

File: Mathlib/Data/Nat/Digits.lean [hunk 4]
```diff
@@ -516,6 +517,59 @@ lemma self_div_pow_eq_ofDigits_drop (i n : ℕ) (h : 2 ≤ p):
...
+        nth_rw 2 [← one_mul <| ofDigits p tl]
+        rw [← add_mul, one_eq_succ_zero, Nat.sub_add_cancel <| zero_lt_of_lt h,
+            Nat.add_sub_add_left]
+  · simp [← h, ofDigits_one]
+  · simp [lt_one_iff.mp h]
+    cases L
+    · simp
+    · simp [ofDigits]
+
+theorem sub_one_mul_sum_div_pow_eq_sub_sum_digits (n : ℕ):
```
Comment (ocfnash) [inline Mathlib/Data/Nat/Digits.lean:559]:
```suggestion
theorem sub_one_mul_sum_log_div_pow_eq_sub_sum_digits (n : ℕ) :
```

**Review by ocfnash — COMMENTED** [review]
Just some minor suggestions.

File: Mathlib/Data/Nat/Digits.lean (hunk 2 above)
Comment (mo271) [reply Mathlib/Data/Nat/Digits.lean:524]: sure, why not.
I had taken the formulation from 
https://github.com/leanprover-community/mathlib4/blob/f2f99ebf8c9a9441e79cd57418dc81a9777c217d/Mathlib/Data/Nat/Digits.lean#L227-L228

File: Mathlib/Data/Nat/Digits.lean [hunk 5]
```diff
@@ -516,6 +517,56 @@ lemma self_div_pow_eq_ofDigits_drop (i n : ℕ) (h : 2 ≤ p):
...
+  · simp [ofDigits_one]
+  · simp [lt_one_iff.mp h]
+    cases L
+    · simp
+    · simp [ofDigits]
+
+theorem sub_one_mul_sum_log_div_pow_eq_sub_sum_digits (n : ℕ) :
+     (p - 1) * ∑ i in range (log p n).succ, n / p ^ i.succ = n - (p.digits n).sum := by
+  obtain h | rfl | h : 1 < p ∨ 1 = p ∨ p < 1 := trichotomous 1 p
+  · by_cases hn : n ≠ 0
```
Comment (ocfnash) [inline Mathlib/Data/Nat/Digits.lean:559]: If you like:
```suggestion
  · rcases eq_or_ne n 0 with rfl | hn
```

**Review by ocfnash — APPROVED** [review]
I think the proofs could still be golfed but the statements and names are now good so it's not so important.

One suggestion.

bors d+

**Pull request closed without merging**